src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11HKDF.java

```diff
@@ -167,8 +168,15 @@ private <T> T derive(String alg, AlgorithmParameterSpec derivationSpec,
         checkDerivedKeyType(ki, alg);
         P11KeyGenerator.checkKeySize(ki.keyGenMech, outLen * 8, token);
 
-        P11Key p11BaseKey = convertKey(baseKey, (isExtract ? "IKM" : "PRK") +
-                " could not be converted to a token key for HKDF derivation.");
+        P11Key p11BaseKey = null;
+        try {
+            p11BaseKey = convertKey(baseKey, (isExtract ? "IKM" : "PRK") +
+                    " could not be converted to a token key for HKDF derivation.");
+        } catch (ProviderException pe) {
+            if (!isExtract || !"RAW".equalsIgnoreCase(baseKey.getFormat())) {
+                throw pe;
+            }
```

Comment:
How about merging line 202-207 with this code block, so that the key/data conversion is together.
```
+        long baseKeyID = 0L;
+        P11Key p11BaseKey = null;
+        try {
+            p11BaseKey = convertKey(baseKey, (isExtract ? "IKM" : "PRK") +
+                    " could not be converted to a token key for HKDF derivation.");
+            baseKeyID = p11BaseKey.getKeyID();
+        } catch (ProviderException pe) {
+            // special handling for FIPS mode when key cannot be imported
+            if (isExtract && "RAW".equalsIgnoreCase(baseKey.getFormat())) {
+                baseKeyID = convertKeyToData(baseKey);
+            } else {
+                throw pe;
+            }
+        }
```